tools/tinygobb/builder/builder.go

```diff
@@ -0,0 +1,328 @@
+	jobs        []BuildJob       // list of jobs that will be added to the jobQueue
+	results     []BuildResult    // list of received build results
+	errors      []BuildError     // list of received build errors
+	logger      *log.Logger      // optional, user provided logger
+}
+
+// Generate a new Builder struct with a configuration.
+func NewBuilder(worker uint) (Builder, error) {
+	if worker == 0 {
+		return Builder{}, ErrNoWorkers
+	}
+
+	return Builder{
+		jobQueue:    make(chan BuildJob, worker),
+		resultQueue: make(chan BuildResult, 0xFF),
+		errQueue:    make(chan BuildError, 0xFF),
+		worker:      worker,
+		state:       Setup,
+		jobs:        make([]BuildJob, 0),
+		results:     make([]BuildResult, 0),
+		errors:      make([]BuildError, 0),
+		logger:      log.Default(),
+	}, nil
+}
+
+// Add a new job to the build queue, as long as the builder is till in the Setup state.
+func (b *Builder) AddJob(job BuildJob) error {
+	if b.state != Setup {
+		return fmt.Errorf("job: builder in invalid state, cannot add new job")
+	}
+
+	if job.goPkgPath == "" {
+		return fmt.Errorf("job: invalid gopkg, cannot add '%v'", job.goPkgPath)
+	}
+
+	b.jobs = append(b.jobs, job)
+	return nil
+}
+
+// Provide a custom logger for the builder. By default, the logger will be turned off.
+func (b *Builder) SetLogger(logger *log.Logger) {
+	b.logger = logger
+}
+
+// The worker function that will be run by the Builder.
+// The job queue as well as the result and error queue are provided by the builder.
+// TODO: make the channels unidirectional?
+// TODO: how about the mutex syncing for the logger? maybe just leave it out for now
+// func worker(jobq chan BuildJob, resultq chan BuildResult, errq chan BuildError, id uint, logger *log.Logger) {
+func worker(jobq chan BuildJob, resultq chan BuildResult, errq chan BuildError, id uint) {
+	fmt.Printf("[%d] spawned worker %d\n", id, id)
+	for {
+		job, ok := <-jobq
+		if !ok {
+			fmt.Printf("[%d] finish worker %d\n", id, id)
+			return
+		}
+
+		var errBuf bytes.Buffer
+
+		// run the build process
+		fields := strings.Fields(job.buildCommand)
+		fields = append(fields, "-o", *job.buildDir)
+
+		c := exec.Command(fields[0], fields[1:]...)
+		c.Env = append(os.Environ(), job.env...)
+		c.Dir = job.goPkgPath
+		c.Stderr = &errBuf
+
+		fmt.Printf("[%d] running '%v'\n", id, c)
+		t0 := time.Now()
+		err := c.Run()
+		t1 := time.Now()
+
+		if err != nil {
+			// the subprocess failed with a non-zero exit code, so create BuildError
+			fmt.Printf("[%d] error building %v:%v", id, job.goPkgPath, errBuf.String())
+			errq <- BuildError{
+				buildJob: job,
+				buildErr: errBuf.String(),
+			}
+		} else {
+			fmt.Printf("[%d] built %v\n", id, job.goPkgPath)
+			pkgNameToken := strings.Split(job.goPkgPath, "/")
+			binPath := job.goPkgPath + "/" + pkgNameToken[len(pkgNameToken)-1]
+
+			f, err := os.Stat(binPath)
+			if err != nil {
+				fmt.Printf("worker: could not find file %v\n", binPath)
+			}
+
+			resultq <- BuildResult{
```

Comment:
suggstion: you could accumulate all that printing stuff in a bytes.Buffer and return it in the BuildError and BuildResult? that way we don't get noise all over the stdout/stderr.

Also, this error case os.Stat at line 228? should that be on your errq or resultq?